File: cpp/questions/climbStairs.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
private:
    std::vector<int> cc;
public:
    int climbStairs(int n) {
        if (cc.size() < n+1) {
            cc.resize(n+1, -1);
        }
        if (cc[n] == -1) {
            if (n <= 1) {
                cc[n] = 1;
            } else {
                cc[n] = climbStairs(n-1) + climbStairs(n-2);
            }
        }
        return cc[n];
    }
};

File: python/questions/climbStairs.py
class Solution:
    def climbStairs(self, n: int) -> int:
        cc = [-1] * (n+1)
        def cs(n: int) -> int:
            nonlocal cc
            if cc[n] != -1:
                return cc[n]
            if n <= 1:
                cc[n] = 1
            else:
                cc[n] = cs(n-1) + cs(n-2)
            return cc[n]
        return cs(n)
